### DISCUSSION TITLE
Server (linux) integrated into golang (cgo) compile error

### DISCUSSION THREAD

**Discussion by mark2021blue** [trustwallet/wallet-core#2234] [Q&A]
**Describe the bug**
Server (linux) integrated into golang (cgo) compile error

TrustWalletCore library version: v2.8.0， based on the source code docker build
Project environment OS：CentOS Linux release 8.3.2011
Golang version: go version go1.18.1 linux/amd64

In file included from /root/project/include/TrustWalletCore/TWCoinType.h:9,
                 from /root/project/include/TrustWalletCore/TWHDWallet.h:10,
                 from service/service.go:6:
/root/project/include/TrustWalletCore/TWBase.h:50:18: error: missing binary operator before token "("
 #if __has_feature(assume_nonnull)
                  ^
/root/project/include/TrustWalletCore/TWBase.h:59:19: error: missing binary operator before token "("
 #if !__has_feature(nullability)

**Comment by mark2021blue** [comment]
Make cmd:
GO111MODULE=on GOOS=linux CGO_ENABLED=1 go build -a -v -mod=vendor ${GCFLAGS} ${LDFLAGS} -o build/${APPLICATION} ${CMD}

**Comment by hewigovens** [comment]
can you check the Dockerfile? it looks like it's missing clang to build wallet core first

**Comment by mark2021blue** [comment]
In the Source code (zip) package used by the dockerfile, no modification

**Comment by hewigovens** [comment]
yeah, you need to build wallet core first to use it in golang

**Comment by mark2021blue** [comment]
yeah,I have built these 3 static library files needed by cgo with dockerfile
-rw-r--r--  1 root root  61280502 May 15 03:54 libprotobuf.a
-rw-r--r--  1 root root 226497742 May 15 03:57 libTrustWalletCore.a
-rw-r--r-- 1 root root 1245008 May 15 03:54 libTrezorCrypto.a

**Comment by hewigovens** [comment]
Can you check if you have installed `libc++1` and `libc++abi1` on host CentOS? and `// #cgo CFLAGS / LDFLAGS` in service.go?

**Comment by mark2021blue** [comment]
#cgo in the service.go file
/*
#cgo CFLAGS: -I/root/project/TrustWalletCore-v2.8.0-linux-for-go/include
#cgo LDFLAGS: -L/root/project/TrustWalletCore-v2.8.0-linux-for-go -lTrustWalletCore -lprotobuf -lTrezorCrypto -lstdc++ -lm
#include <TrustWalletCore/TWHDWallet.h>
#include <TrustWalletCore/TWString.h>
#include <TrustWalletCore/TWData.h>
#include <TrustWalletCore/TWPrivateKey.h>
#include <TrustWalletCore/TWPublicKey.h>
#include <TrustWalletCore/TWMnemonic.h>
*/
import "C"

These two files are not found in the system, how to install on centos? Can you give me a hint? Thanks!
[root@dev project]# find / -name  "\*libc++1*"
[root@dev project]# find / -name "\*libc++abi1*"

**Comment by mark2021blue** [comment]
I found these two files in the docker environment (Ubuntu), I wonder if they can be copied to the centos environment for use? Or does it need to be reinstalled in the centos environment?

root@35ae9c590d9d:/wallet-core/# find / -name "*libc++abi1*"
/var/lib/dpkg/info/libc++abi1-10:amd64.md5sums
/var/lib/dpkg/info/libc++abi1-10:amd64.triggers
/var/lib/dpkg/info/libc++abi1-10:amd64.shlibs
/var/lib/dpkg/info/libc++abi1-10:amd64.list
/usr/share/lintian/overrides/libc++abi1-10
/usr/share/doc/libc++abi1-10

root@35ae9c590d9d:/wallet-core# find / -name "*libc++1*"
/var/lib/dpkg/info/libc++1-10:amd64.shlibs
/var/lib/dpkg/info/libc++1-10:amd64.triggers
/var/lib/dpkg/info/libc++1-10:amd64.list
/var/lib/dpkg/info/libc++1-10:amd64.md5sums
/usr/share/lintian/overrides/libc++1-10
/usr/share/doc/libc++1-10

**Comment by mark2021blue** [comment]
Try the following installation, the installation did not report an error, but after installation, the two words were not found.

[root@dev project]# yum -y install clang
Failed to set locale, defaulting to C
Last metadata expiration check: 0:48:20 ago on Tue May 17 11:38:57 2022.
Dependencies resolved.
========================================================================================================================================================================================================================================================================================================================================================================
 Package                                                                                       Architecture                                                               Version                                                                                                   Repository                                                                     Size
========================================================================================================================================================================================================================================================================================================================================================================
Installing:
 clang                                                                                         x86_64                                                                     12.0.1-4.module_el8.5.0+1025+93159d6c                                                                     appstream                                                                      82 k
Installing dependencies:
 clang-libs                                                                                    x86_64                                                                     12.0.1-4.module_el8.5.0+1025+93159d6c                                                                     appstream                                                                      22 M
 clang-resource-filesystem                                                                     x86_64                                                                     12.0.1-4.module_el8.5.0+1025+93159d6c                                                                     appstream                                                                      14 k
 llvm-libs                                                                                     x86_64                                                                     12.0.1-2.module_el8.5.0+918+ed335b90                                                                      appstream                                                                      23 M
Installing weak dependencies:
 compiler-rt                                                                                   x86_64                                                                     12.0.1-1.module_el8.5.0+892+54d791e1                                                                      appstream                                                                     3.8 M
 libomp                                                                                        x86_64                                                                     12.0.1-1.module_el8.5.0+892+54d791e1                                                                      appstream                                                                     343 k
 libomp-devel                                                                                  x86_64                                                                     12.0.1-1.module_el8.5.0+892+54d791e1                                                                      appstream                                                                      24 k
Enabling module streams:
 llvm-toolset                                                                                                                                                             rhel8                                                                                                                                                                                        

Transaction Summary
========================================================================================================================================================================================================================================================================================================================================================================
Install  7 Packages

Total download size: 49 M
Installed size: 213 M
Downloading Packages:
(1/7): clang-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64.rpm                                                                                                                                                                                                                                                                           2.0 MB/s |  82 kB     00:00    
(2/7): clang-resource-filesystem-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64.rpm                                                                                                                                                                                                                                                       352 kB/s |  14 kB     00:00    
(3/7): libomp-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64.rpm                                                                                                                                                                                                                                                                            12 MB/s | 343 kB     00:00    
(4/7): libomp-devel-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64.rpm                                                                                                                                                                                                                                                                     2.5 MB/s |  24 kB     00:00    
(5/7): compiler-rt-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64.rpm                                                                                                                                                                                                                                                                       15 MB/s | 3.8 MB     00:00    
(6/7): clang-libs-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64.rpm                                                                                                                                                                                                                                                                       16 MB/s |  22 MB     00:01    
(7/7): llvm-libs-12.0.1-2.module_el8.5.0+918+ed335b90.x86_64.rpm                                                                                                                                                                                                                                                                         16 MB/s |  23 MB     00:01    
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                                                                                                                                                                                                    33 MB/s |  49 MB     00:01     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                                                                                                                                                                                                                1/1 
  Installing       : clang-resource-filesystem-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64                                                                                                                                                                                                                                                                         1/7 
  Installing       : llvm-libs-12.0.1-2.module_el8.5.0+918+ed335b90.x86_64                                                                                                                                                                                                                                                                                          2/7 
  Installing       : compiler-rt-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64                                                                                                                                                                                                                                                                                        3/7 
  Installing       : libomp-devel-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64                                                                                                                                                                                                                                                                                       4/7 
  Installing       : libomp-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64                                                                                                                                                                                                                                                                                             5/7 
  Installing       : clang-libs-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64                                                                                                                                                                                                                                                                                        6/7 
  Installing       : clang-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64                                                                                                                                                                                                                                                                                             7/7 
  Running scriptlet: clang-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64                                                                                                                                                                                                                                                                                             7/7 
  Verifying        : clang-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64                                                                                                                                                                                                                                                                                             1/7 
  Verifying        : clang-libs-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64                                                                                                                                                                                                                                                                                        2/7 
  Verifying        : clang-resource-filesystem-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64                                                                                                                                                                                                                                                                         3/7 
  Verifying        : compiler-rt-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64                                                                                                                                                                                                                                                                                        4/7 
  Verifying        : libomp-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64                                                                                                                                                                                                                                                                                             5/7 
  Verifying        : libomp-devel-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64                                                                                                                                                                                                                                                                                       6/7 
  Verifying        : llvm-libs-12.0.1-2.module_el8.5.0+918+ed335b90.x86_64                                                                                                                                                                                                                                                                                          7/7 
Installed products updated.

Installed:
  clang-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64      clang-libs-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64   clang-resource-filesystem-12.0.1-4.module_el8.5.0+1025+93159d6c.x86_64   compiler-rt-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64   libomp-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64   libomp-devel-12.0.1-1.module_el8.5.0+892+54d791e1.x86_64  
  llvm-libs-12.0.1-2.module_el8.5.0+918+ed335b90.x86_64  

Complete!

**Comment by hewigovens** [comment]
They're called `libcxx` and `libcxxabi` on CentOS

**Comment by mark2021blue** [comment]
I have tried the following methods, but the installation was not successful. Due to time constraints, I have to give up the cgo method for the time being and use the golang shell to call the C module.

https://stackoverflow.com/questions/39332406/install-libc-on-ubuntu
https://releases.llvm.org/14.0.0/projects/libcxx/docs/BuildingLibcxx.html#the-default-build
https://releases.llvm.org/download.html#6.0.0

**Comment by mark2021blue** [comment]
https://developer.trustwallet.com/wallet-core/integration-guide/server-side

You might want to copy and run main outside of the docker container, make sure you have libc++1 and libc++abi1 installed in your host Ubuntu.

I compiled the sample (golang version) from the docker environment and copied it to the centos environment. It works normally. In the centos environment, the dependencies of the cgo-dependent library "-lTrustWalletCore -lprotobuf -lTrezorCrypto -lstdc++ -lm" are normal, but the compilation time this error,

[root@ project]# ldd main
        linux-vdso.so.1 (0x00007ffe8f5bd000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f7c6536b000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f7c64fe4000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f7c64dc5000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f7c64bad000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f7c647f4000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f7c65323000)

[root@project]# ./main
==> calling wallet core from go
Bitcoin wallet: 
         address: bc1qw29x4hrt6tahz4jvuhzrq6y5el3spqt499zuay

**Comment by mark2021blue** [comment]
GO111MODULE=on GOOS=linux CGO_ENABLED=1 CC="/usr/bin/clang-12" CXX="/usr/bin/clang++-12" go build -a -v -mod=vendor ${GCFLAGS} ${LDFLAGS} -o build/${APPLICATION} ${CMD}

I tried to specify CC and CXX to clang in the makefile, the error "#if __has_feature(assume_nonnull)" was not reported, but the following error was reported，looks like a symbol problem :

/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Signer.cpp.o): In function `std::basic_string_view<char, std::char_traits<char> > google::protobuf::internal::TransparentSupport<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::ImplicitConvert<void>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/wallet-core/build/local/include/google/protobuf/map.h:225: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator std::basic_string_view<char, std::char_traits<char> >() const'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Transaction.cpp.o): In function `basic_pointerbuf':
/usr/include/boost/detail/basic_pointerbuf.hpp:42: undefined reference to `std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::basic_stringbuf()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Signer.cpp.o): In function `to_hex[abi:cxx11](long)':
/wallet-core/src/Icon/Signer.cpp:26: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Signer.cpp.o): In function `TW::Icon::Signer::sign() const':
/wallet-core/src/Icon/Signer.cpp:91: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Transaction.cpp.o): In function `nlohmann::detail::lexer<nlohmann::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, unsigned long, double, std::allocator, nlohmann::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char> > >, nlohmann::detail::iterator_input_adapter<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::scan_number()':
/wallet-core/build/local/include/nlohmann/detail/input/lexer.hpp:1245: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/wallet-core/build/local/include/nlohmann/detail/input/lexer.hpp:1248: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/wallet-core/build/local/include/nlohmann/detail/input/lexer.hpp:1261: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/wallet-core/build/local/include/nlohmann/detail/input/lexer.hpp:1264: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Transaction.cpp.o):/wallet-core/build/local/include/nlohmann/detail/input/lexer.hpp:1278: more undefined references to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()' follow
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Serialization.cpp.o): In function `typeUrl[abi:cxx11](protocol::Transaction_Contract_ContractType)':
/wallet-core/src/Tron/Serialization.cpp:24: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(extension_set.cc.o): In function `bool google::protobuf::internal::ParseMessageSetItemImpl<google::protobuf::internal::ExtensionSet::ParseMessageSetItemLite(google::protobuf::io::CodedInputStream*, google::protobuf::internal::ExtensionFinder*, google::protobuf::internal::FieldSkipper*)::MSLite>(google::protobuf::io::CodedInputStream*, google::protobuf::internal::ExtensionSet::ParseMessageSetItemLite(google::protobuf::io::CodedInputStream*, google::protobuf::internal::ExtensionFinder*, google::protobuf::internal::FieldSkipper*)::MSLite)':
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/wire_format_lite.h:1864: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(extension_set_heavy.cc.o): In function `bool google::protobuf::internal::ParseMessageSetItemImpl<google::protobuf::internal::ExtensionSet::ParseMessageSetItem(google::protobuf::io::CodedInputStream*, google::protobuf::internal::ExtensionFinder*, google::protobuf::internal::MessageSetFieldSkipper*)::MSFull>(google::protobuf::io::CodedInputStream*, google::protobuf::internal::ExtensionSet::ParseMessageSetItem(google::protobuf::io::CodedInputStream*, google::protobuf::internal::ExtensionFinder*, google::protobuf::internal::MessageSetFieldSkipper*)::MSFull)':
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/wire_format_lite.h:1864: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(common.cc.o): In function `google::protobuf::internal::LogMessage::operator<<(google::protobuf::uint128 const&)':
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/common.cc:204: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(int128.cc.o): In function `google::protobuf::operator<<(std::ostream&, google::protobuf::uint128 const&)':
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/int128.cc:153: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(strutil.cc.o): In function `google::protobuf::StrAppend(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, google::protobuf::strings::AlphaNum const&)':
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/strutil.cc:1574: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(strutil.cc.o): In function `google::protobuf::StrAppend(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, google::protobuf::strings::AlphaNum const&, google::protobuf::strings::AlphaNum const&)':
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/strutil.cc:1579: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/strutil.cc:1580: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(strutil.cc.o): In function `google::protobuf::StrAppend(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, google::protobuf::strings::AlphaNum const&, google::protobuf::strings::AlphaNum const&, google::protobuf::strings::AlphaNum const&)':
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/strutil.cc:1590: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/strutil.cc:1591: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libprotobuf.a(strutil.cc.o):/wallet-core/build/local/src/protobuf/protobuf-3.19.2/src/google/protobuf/stubs/strutil.cc:1592: more undefined references to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()' follow
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(TWData.cpp.o): In function `boost::enable_if<boost::is_output_streamable<char, char, std::char_traits<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::type boost::to_string<char>(char const&)':
/usr/include/boost/exception/to_string.hpp:63: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(TWData.cpp.o): In function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > boost::exception_detail::object_hex_dump<char>(char const&, unsigned long)':
/usr/include/boost/exception/detail/object_hex_dump.hpp:39: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(TWData.cpp.o): In function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > boost::exception_detail::object_hex_dump<boost::error_info<boost::algorithm::bad_char_, char> >(boost::error_info<boost::algorithm::bad_char_, char> const&, unsigned long)':
/usr/include/boost/exception/detail/object_hex_dump.hpp:39: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(TWData.cpp.o): In function `boost::exception_detail::error_info_container_impl::diagnostic_information(char const*) const':
/usr/include/boost/exception/info.hpp:102: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Base64.cpp.o): In function `TW::Base64::convertFromBase64Url(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
/wallet-core/src/Base64.cpp:38: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Base64.cpp.o): In function `TW::Base64::convertToBase64Url(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
/wallet-core/src/Base64.cpp:54: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Cbor.cpp.o): In function `TW::Cbor::Decode::dumpToStringInternal[abi:cxx11]() const':
/wallet-core/src/Cbor.cpp:392: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Cbor.cpp.o): In function `TW::Cbor::Decode::dumpToString[abi:cxx11]() const':
/wallet-core/src/Cbor.cpp:464: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Asset.cpp.o): In function `TW::EOS::Asset::fromString(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)':
/wallet-core/src/EOS/Asset.cpp:70: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/wallet-core/src/EOS/Asset.cpp:71: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
/root/project/TrustWalletCore-v2.8.0-linux-for-go/libTrustWalletCore.a(Address.cpp.o): In function `TW::Nimiq::Address::isValid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/wallet-core/src/Nimiq/Address.cpp:41: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data()'
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:23: linux] Error 2

**Comment by hewigovens** [comment]
can you build your app inside docker container? or try link `-lcxx`

**Comment by mark2021blue** [comment]
If my application is built in docker, it is a little inconvenient to deploy

try link -lcxx

make cmd:
GO111MODULE=on GOOS=linux CGO_ENABLED=1 CXX="/usr/bin/clang++-12" CC="/usr/bin/clang-12" go build -a -v -mod=vendor ${GCFLAGS} ${LDFLAGS } -o build/${APPLICATION} ${CMD}

cgo:
/*
#cgo CFLAGS: -I/root/project/TrustWalletCore-v2.8.0-linux-for-go/include
#cgo LDFLAGS: -L/root/project/TrustWalletCore-v2.8.0-linux-for-go -lTrustWalletCore -lprotobuf -lTrezorCrypto -lcxx -lm
#include <TrustWalletCore/TWHDWallet.h>
#include <TrustWalletCore/TWString.h>
#include <TrustWalletCore/TWData.h>
#include <TrustWalletCore/TWPrivateKey.h>
#include <TrustWalletCore/TWPublicKey.h>
#include <TrustWalletCore/TWMnemonic.h>
*/
import "C"

Report not found library error:
/usr/bin/ld: cannot find -lcxx
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:23: linux] Error 2

**Comment by hewigovens** [comment]
not familiar with CentOS, if `-lcxx` doesn't work, how about `-lc++`?

**Comment by mark2021blue** [comment]
have already tried...
Thank you very much for your assistance!
I'll keep watching for a solution to this problem...
